### ISSUE TITLE
Enabling lanzaboote on initial installation

### ISSUE THREAD

**Issue by Ten0** [nix-community/lanzaboote#397]
Readme specifies:
https://github.com/nix-community/lanzaboote/blob/f5a3a7dff44d131807fc1a89fbd8576cd870334a/docs/QUICK_START.md?plain=1#L33-L35

I'd like to be able to apply a configuration with lanzaboote enabled as I run `nixos-install`, and not have to manually temporarily disable it from my config during the installation, then re-enable it after the installation.

Could you please expand on the limitations that prevent from running `sbctl create-keys` prior to installation, then applying the initial configuration with lanzaboote enabled right away?

**Comment by arthsmn** [comment]
You can create your own ISO with your keys in it and a signed kernel. Then, you would only need to import the keys into the machine and copy the keys from the ISO to the installation.

**Comment by Ten0** [comment]
Hey, thanks for your answer.
I didn't mean that I wanted secure boot to be enabled prior to installation. I meant that the guide seems to express that applying a configuration that has lanzaboote enabled instead of systemd boot during initial `nixos-install` wouldn't work, and I was wondering why creating keys in `/mnt/etc/secureboot` prior to installation wouldn't work (or if it would, why the guide seems to suggest otherwise).
Thanks!

**Comment by arthsmn** [comment]
> Hey, thanks for your answer.
> I didn't mean that I wanted secure boot to be enabled prior to installation. I meant that the guide seems to express that applying a configuration that has lanzaboote enabled instead of systemd boot during initial `nixos-install` wouldn't work, and I was wondering why creating keys in `/mnt/etc/secureboot` prior to installation wouldn't work (or if it would, why the guide seems to suggest otherwise).
> Thanks!

I think this is a misunderstanding of the guide, as I'm pretty sure I installed lanzaboote directly. As the module just replaces systemd-boot, in theory if it has everything in place it should work.

**Comment by RaitoBezarius** [comment]
Hi there, `sbctl create-keys` is frowned upon in activation because this is a side effect that creates private keys and private key creation in an activation script is a bit icky.

Nonetheless, as long as you copied the keys from another location, you can immediately use lanzaboote on the first nixos-install. Lanzaboote will not enforce Secure Boot if Secure Boot is disabled, as soon as you enroll the keys, Secure Boot will be enabled and your bootables will be signed according to the right keys, so it should work out fine in the end.

The guide is a bit conservative to avoid bad experiences, but it could indeed benefit from some clarification.

**Comment by kuflierl** [comment]
> Hello! I've been using this successfully for a few months, and I'd like to generalize its usage.
> 
> Readme specifies:
> 
> https://github.com/nix-community/lanzaboote/blob/f5a3a7dff44d131807fc1a89fbd8576cd870334a/docs/QUICK_START.md?plain=1#L33-L35
> 
> I'd like to be able to apply a configuration with lanzaboote enabled as I run `nixos-install`, and not have to manually temporarily disable it from my config during the installation, then re-enable it after the installation.
> 
> Could you please expand on the limitations that prevent from running `sbctl create-keys` prior to installation, then applying the initial configuration with lanzaboote enabled right away? Thanks!

This is the whole point of #384. This subject is kinda controversial with the maintainers/developers of the repository as seen in the comments of the PR.

**Comment by Ten0** [comment]
> This is the whole point of https://github.com/nix-community/lanzaboote/pull/384

As far as I'm concerned I'm fine with needing to run one or two extra commands as part of my install. That is already scripted anyway for all partitioning, encryption, git clone, hardware config generation, etc.

So I don't mind that it doesn't run `create-keys`. I'm mostly just scared that it's specified that I should apply a different config with systemd-boot instead before first boot, so I'm wondering if perhaps there's some magic in the install script that checks whether systemd-uboot is the bootloader, and that won't work with lanzaboote.

At this point I'm rather reassured by https://github.com/nix-community/lanzaboote/issues/397#issuecomment-2423698765 but I haven't tested it yet.

**Comment by qm3ster** [comment]
I did `sudo sbctl create-keys -e /mnt/var/lib/sbctl --disable-landlock`
but `sudo nixos-install -v --flake /mnt/etc/nixos#nook` fails with:
```sh
evaluating file '<nix/derivation-internal.nix>'
installing the boot loader...
setting up /etc...
Installing Lanzaboote to "/boot"...
Failed to install generation 1: Get stub name: No such file or directory (os error 2)
Failed to install bootloader
```
which localizes the issue to this call: https://github.com/nix-community/lanzaboote/blob/2e425f3da6ce7f5b34fa6eaf7a2a7f78dbabcc85/rust/tool/systemd/src/install.rs#L267

Pls help me figure it out.

**Comment by andsens** [comment]
> Hope this helps someone.

It did! Thank you :-)

> Funnily enough after everything worked out I didn't get a boot entry added to the EFI flash so I can't actually try booting yet, but I have to go to bed now.

That's because lanzaboote doesn't do that (yet?). It piggybacks on the sd-boot EFI entry that was installed before the user switched (I think). Here's a workaround:

```
mv /mnt/boot/EFI/BOOT/BOOTX64.EFI /mnt/boot/EFI/BOOT/BOOTX64.EFI.BAK
mv /mnt/boot/EFI/systemd/systemd-bootx64.efi /mnt/boot/EFI/systemd/systemd-bootx64.efi.bak
bootctl --esp-path /mnt/boot install
mv /mnt/boot/EFI/BOOT/BOOTX64.EFI.BAK /mnt/boot/EFI/BOOT/BOOTX64.EFI
mv /mnt/boot/EFI/systemd/systemd-bootx64.efi.bak /mnt/boot/EFI/systemd/systemd-bootx64.efi
```

`bootctl install` both sets up the EFI entry, but also copies over the unsigned systemd-boot EFI app. So to get just the EFI entry part we move the signed binaries out of the way, run `bootctl`, and then move them back into place.

**Comment by Ten0** [comment]
So IIUC the above PR still needs an intermediate reboot before SB is enabled?
I'm curious why but couldn't understand from the documentation added in the PR (or missed it).
From [the comments above](https://github.com/nix-community/lanzaboote/issues/397#issuecomment-3280191884) it seems the only thing that was missing was lanzaboote adding the appropriate EFI entry upon install, like systemd-boot would? 
So it's unclear to me why the PR seems to instead be about auto creating keys and stuff.

**Comment by andsens** [comment]
I agree with @Ten0. The enrollment on initial boot is a great addition, but it has nothing to do with setting up the initial EFI entry.  
@nikstur, would you kindly reopen this?

**Comment by nikstur** [comment]
> So IIUC the above PR still needs an intermediate reboot before SB is enabled?

A reboot is always required because we leverage systemd-boot to enroll the keys (which does that via a reboot).

This way, you can use Lanzaboote in self-enrolling appliances, i.e. create an image completely without keys, start it somewhere and it enrolls itself into Secure Boot. Doing this at install time is possible but IMO not desirable. They way I have implemented it now, it works for more use-cases: "normal" physical machines, image-based physical machines, VM images, tests etc.

> So it's unclear to me why the PR seems to instead be about auto creating keys and stuff.

To automatically provision a system, you need to create the keys and enroll them. The solutions presented in this issue were a possible implementation but ultimately not what we wanted to follow.

You can already our implementation to auto provision an entire machine. This is proven by our tests and is possible because of the EFI fallback path that is booted if no other entry is bootable.

> The enrollment on initial boot is a great addition, but it has nothing to do with setting up the initial EFI entry.

Setting the EFI entry is just one part of the provisioning steps that can be done but doesn't necessarily need to be done in all situations. Automatically setting the entry is covered in #126.

In summary: I view "Enabling lanzaboote on initial install" as implemented and a possible follow up is in #126. Thus, not re-opening.
